airbyte-workers/src/main/java/io/airbyte/workers/temporal/PartialSuccessTemporalAttemptExecution.java

```diff
@@ -0,0 +1,221 @@
+                                         Predicate<OUTPUT> shouldAttemptAgainPredicate,
+                                         BiFunction<INPUT, OUTPUT, INPUT> computeNextAttemptInputFunction,
+                                         int maxRetriesCount,
+                                         TemporalAttemptExecutionFactory<INPUT, OUTPUT> temporalAttemptExecutionFactory) {
+    this.workspaceRoot = workspaceRoot;
+    this.jobRunConfig = jobRunConfig;
+    this.workerSupplier = workerSupplier;
+    this.inputSupplier = initialInputSupplier;
+    this.cancellationHandler = cancellationHandler;
+    this.shouldAttemptAgainPredicate = shouldAttemptAgainPredicate;
+    this.computeNextAttemptInputFunction = computeNextAttemptInputFunction;
+    this.maxRetriesCount = maxRetriesCount;
+    this.temporalAttemptExecutionFactory = temporalAttemptExecutionFactory;
+  }
+
+  @Override
+  public List<OUTPUT> get() {
+    INPUT input = inputSupplier.get();
+    final AtomicReference<OUTPUT> lastOutput = new AtomicReference<>();
+    List<OUTPUT> outputCollector = new ArrayList<>();
+
+    int i = 0;
+    while (true) {
+      if (i >= maxRetriesCount) {
+        LOGGER.info("Max retries reached: {}", i);
+        break;
+      }
+
+      final boolean hasLastOutput = lastOutput.get() != null;
+      final boolean shouldAttemptAgain = !hasLastOutput || shouldAttemptAgainPredicate.test(lastOutput.get());
+      LOGGER.info("Last output present: {}. Should attempt again: {}", lastOutput.get() != null, shouldAttemptAgain);
+      if (hasLastOutput && !shouldAttemptAgain) {
+        break;
+      }
+
+      LOGGER.info("Starting attempt: {} of {}", i, maxRetriesCount);
+
+      Supplier<INPUT> resolvedInputSupplier = !hasLastOutput ? inputSupplier : () -> computeNextAttemptInputFunction.apply(input, lastOutput.get());
+
+      final TemporalAttemptExecution<INPUT, OUTPUT> temporalAttemptExecution = temporalAttemptExecutionFactory.create(
+          workspaceRoot,
+          jobRunConfig,
+          workerSupplier,
+          resolvedInputSupplier,
+          cancellationHandler);
+      lastOutput.set(temporalAttemptExecution.get());
+      outputCollector.add(lastOutput.get());
+      i++;
+    }
+
+    return outputCollector;
+  }
+
+  // interface to make testing easier.
+  @FunctionalInterface
+  interface TemporalAttemptExecutionFactory<INPUT, OUTPUT> {
+
+    TemporalAttemptExecution<INPUT, OUTPUT> create(Path workspaceRoot,
+                                                   JobRunConfig jobRunConfig,
+                                                   CheckedSupplier<Worker<INPUT, OUTPUT>, Exception> workerSupplier,
+                                                   Supplier<INPUT> inputSupplier,
+                                                   CancellationHandler cancellationHandler);
+
+  }
+
+  private static class TestTemporalAttemptExecutionFactory<INPUT, OUTPUT> implements TemporalAttemptExecutionFactory<INPUT, OUTPUT> {
```

Comment:
Is it meant to be used in tests? If so then it shouldn't live in `src` neither should the @visible4testing ctor.

Reply:
same comment from me

Reply:
agreed.